### DISCUSSION TITLE
Attention Coefficient Matrix conversion

### DISCUSSION THREAD

**Discussion by tsjshg** [danielegrattarola/spektral#404] [Q&A] [answered]
In `GATConv` layer, if we set  `return_attn_coef` argument `True` we can get the attention coefficient matrix with each head. But it does not have the shape `n_nodes` x `n_nodes`, I convert it to the squared matrix form with the following code. Could you please check whether these are correct or not?

`a` is adjacency matrix as  SparseTensor type
`att` is the returned attention coefficient matrix
`n_nodes` is the number of nodes as int

When `add_self_loops` argument is `True`
```python
from spektral.layers.ops import  add_self_loops_indices

self_looped_idx = add_self_loops_indices(a.indices)
tf.scatter_nd(self_looped_idx, att[..., 0, 0], (n_nodes, n_nodes))
```

When `add_self_loops` argument is `False`
```python
tf.scatter_nd(a.indices, att[..., 0, 0], (n_nodes, n_nodes))
```

I've confirmed them with [CATConv code](https://github.com/danielegrattarola/spektral/blob/5a714303f09f30d93f122c760e7b718f4497768d/spektral/layers/convolutional/gat_conv.py#L11), I think this might be correct...

**Comment by danielegrattarola** [comment, marked as answer]
Yes, this looks correct. 
The attention indices of the GAT layer are mapped 1:1 to the edges, and the self-loops are added at the end of the edge list. 
Were you able to achieve what you wanted to do?

**Reply by tsjshg** [reply]
Thank you for your kind reply.
Yes, I've successfully made my first graph attention autoencoder for the drug repositioning research. Spektral is easy to use, so much helpful for me! This is a great library! Thanks.
